wgpu-core/src/indirect_draw_validation.rs

```diff
@@ -0,0 +1,1080 @@
+    device: &dyn hal::DynDevice,
+    supports_indirect_first_instance: bool,
+) -> Result<Box<dyn hal::DynShaderModule>, CreateDrawIndirectValidationPipelineError> {
+    let src = get_shader_source(supports_indirect_first_instance);
+
+    let module = naga::front::wgsl::parse_str(&src).map_err(|inner| {
+        CreateShaderModuleError::Parsing(naga::error::ShaderError {
+            source: src.clone(),
+            label: None,
+            inner: Box::new(inner),
+        })
+    })?;
+    let info = crate::device::create_validator(
+        wgt::Features::PUSH_CONSTANTS,
+        wgt::DownlevelFlags::empty(),
+        naga::valid::ValidationFlags::all(),
+    )
+    .validate(&module)
+    .map_err(|inner| {
+        CreateShaderModuleError::Validation(naga::error::ShaderError {
+            source: src,
+            label: None,
+            inner: Box::new(inner),
+        })
+    })?;
+    let hal_shader = hal::ShaderInput::Naga(hal::NagaShader {
+        module: std::borrow::Cow::Owned(module),
+        info,
+        debug_source: None,
+    });
+    let hal_desc = hal::ShaderModuleDescriptor {
+        label: None,
+        runtime_checks: wgt::ShaderRuntimeChecks::unchecked(),
+    };
+    let module = unsafe { device.create_shader_module(&hal_desc, hal_shader) }.map_err(
+        |error| match error {
+            hal::ShaderError::Device(error) => {
+                CreateShaderModuleError::Device(DeviceError::from_hal(error))
+            }
+            hal::ShaderError::Compilation(ref msg) => {
+                log::error!("Shader error: {}", msg);
+                CreateShaderModuleError::Generation
+            }
+        },
+    )?;
+
+    Ok(module)
+}
+
+fn create_validation_pipeline(
+    device: &dyn hal::DynDevice,
+    module: &dyn hal::DynShaderModule,
+    pipeline_layout: &dyn hal::DynPipelineLayout,
+) -> Result<Box<dyn hal::DynComputePipeline>, CreateDrawIndirectValidationPipelineError> {
+    let pipeline_desc = hal::ComputePipelineDescriptor {
+        label: None,
+        layout: pipeline_layout,
+        stage: hal::ProgrammableStage {
+            module,
+            entry_point: "main",
+            constants: &Default::default(),
+            zero_initialize_workgroup_memory: false,
+        },
+        cache: None,
+    };
+    let pipeline =
+        unsafe { device.create_compute_pipeline(&pipeline_desc) }.map_err(|err| match err {
+            hal::PipelineError::Device(error) => {
+                CreateComputePipelineError::Device(DeviceError::from_hal(error))
+            }
+            hal::PipelineError::Linkage(_stages, msg) => CreateComputePipelineError::Internal(msg),
+            hal::PipelineError::EntryPoint(_stage) => CreateComputePipelineError::Internal(
+                crate::device::ENTRYPOINT_FAILURE_ERROR.to_string(),
+            ),
+            hal::PipelineError::PipelineConstants(_, error) => {
+                CreateComputePipelineError::PipelineConstants(error)
+            }
```

Comment:
Thought: could these be in some kind of `From` implementation?

Reply:
I don't think we should, device errors in `hal::PipelineError::Device` need to go through `device.handle_hal_error()` unless there is no device yet (in this case). A `From` impl would be easy to misuse.

I think at some point we should coalesce the `Linkage` and `EntryPoint` variants though.